src/validation.cpp

```diff
@@ -3584,13 +3600,20 @@ bool TestBlockValidity(CValidationState& state, const CChainParams& chainparams,
 {
     AssertLockHeld(cs_main);
     assert(pindexPrev && pindexPrev == chainActive.Tip());
-    if (fCheckpointsEnabled && !CheckIndexAgainstCheckpoint(pindexPrev, state, chainparams, block.GetHash()))
+
+    uint256 hash = block.GetHash();
+    if (fCheckpointsEnabled && !CheckIndexAgainstCheckpoint(pindexPrev, state, chainparams, hash))
         return error("%s: CheckIndexAgainstCheckpoint(): %s", __func__, state.GetRejectReason().c_str());
 
+    if (llmq::chainLocksHandler->HasConflictingChainLock(pindexPrev->nHeight + 1, hash)) {
+        return state.DoS(10, error("%s: conflicting with chainlock", __func__), REJECT_INVALID, "bad-chainlock");
+    }
+
     CCoinsViewCache viewNew(pcoinsTip);
     CBlockIndex indexDummy(block);
     indexDummy.pprev = pindexPrev;
     indexDummy.nHeight = pindexPrev->nHeight + 1;
+    indexDummy.phashBlock = &hash;
```

Comment:
While this is ok semantically, this is not ok internal logic wise, so to say, because (as I understand it) `phashBlock` is really meant to point to an item in `mapBlockIndex` (or whatever map is tied to the chain). Pointing it to smth else (even temporary) can break something down the line i.e. you'd expect it to be non-null and thus being a part of the chain but it would not be the case and then all kind of bad things can happen.
```suggestion
```

Reply:
Tried to understand the uses of `phashBlock` and the only reference to hidden/indirect assumptions I found is at https://github.com/dashpay/dash/blob/1db51f569758b3cfdac5b39ad4f908a1261fd0d4/src/validation.cpp#L1969

It's basically a workaround for the same issue I tried to fix with this. As I understand the code that usually sets `phashBlock` to a valid pointer, it's only done this way because "the hash is there, so let's reference it instead of storing a copy". Using a pointer to the key of a map really only looks like an optimization without deeper meaning. IMHO, it's a horrible optimization because it makes the lifecycle/validity of an object dependent on the containing map...that's an invitation for future bugs/crashes.

Reply:
My point is that it's already the case (`phashBlock` is assumed to be null on `CreateNewBlock` in bitcoin codebase), so we should probably stick to that because who knows what backports can bring. Saving few hashing operations is not a good tradeoff for a potential bug in `ConnectBlock` due to different interpretations of `phashBlock == null`.

Reply:
I added a commit on top that skips the test for ChainLock conflicts when there is no `phashBlock`. The reasoning is: If we don't know the hash yet, we also can't check for a conflict because it requires us to know the final hash of the block.

And this also made me better understand the meaning `phashBlock`...if it's unset, it means that the block hash was not solved through PoW yet, so just setting `phashBlock` to a PoW invalid hash is not correct.